src/main/java/org/opentripplanner/apis/gtfs/PatternByServiceDatesFilter.java

```diff
@@ -0,0 +1,109 @@
+   *                       date is defined and this will therefore match all dates.
+   * @param endInclusive The inclusive end date to check the patterns for. If null then no end date
+   *                     is defined this will therefore match all dates.
+   */
+  PatternByServiceDatesFilter(
+    @Nullable LocalDate startInclusive,
+    @Nullable LocalDate endInclusive,
+    Function<Route, Collection<TripPattern>> getPatternsForRoute,
+    Function<Trip, Collection<LocalDate>> getServiceDatesForTrip
+  ) {
+    this.getPatternsForRoute = Objects.requireNonNull(getPatternsForRoute);
+    this.getServiceDatesForTrip = Objects.requireNonNull(getServiceDatesForTrip);
+    // optional
+    this.startInclusive = startInclusive;
+    this.endInclusive = endInclusive;
+
+    if (startInclusive == null && endInclusive == null) {
+      throw new IllegalArgumentException("startInclusive and endInclusive cannot be both null");
+    } else if (
+      startInclusive != null && endInclusive != null && startInclusive.isAfter(endInclusive)
+    ) {
+      throw new IllegalArgumentException("start must be before end");
+    }
+  }
+
+  public PatternByServiceDatesFilter(
+    GraphQLServiceDateFilterInput filterInput,
+    TransitService transitService
+  ) {
+    this(
+      filterInput.getGraphQLStart(),
+      filterInput.getGraphQLEnd(),
+      transitService::getPatternsForRoute,
+      trip -> transitService.getCalendarService().getServiceDatesForServiceId(trip.getServiceId())
+    );
+  }
+
+  /**
+   * Filter the patterns by the service dates that it operates on.
+   */
+  public Collection<TripPattern> filterPatterns(Collection<TripPattern> tripPatterns) {
+    return tripPatterns.stream().filter(this::hasServicesOnDate).toList();
+  }
+
+  /**
+   * Filter the routes by listing all their patterns' service dates and checking if they
+   * operate on the specified dates.
+   */
+  public Collection<Route> filterRoutes(Stream<Route> routeStream) {
+    return routeStream
+      .filter(r -> {
+        var patterns = getPatternsForRoute.apply(r);
+        return !this.filterPatterns(patterns).isEmpty();
+      })
+      .toList();
+  }
+
+  private boolean hasServicesOnDate(TripPattern pattern) {
+    return pattern
+      .scheduledTripsAsStream()
+      .anyMatch(trip -> {
+        var dates = getServiceDatesForTrip.apply(trip);
+
+        return dates
+          .stream()
+          .anyMatch(date ->
+            (
+              startInclusive == null || date.isEqual(startInclusive) || date.isAfter(startInclusive)
+            ) &&
+            (endInclusive == null || date.isEqual(endInclusive) || date.isBefore(endInclusive))
+          );
+      });
+  }
+
+  public static boolean hasServiceDateFilter(GraphQLServiceDateFilterInput serviceDays) {
+    return (
+      serviceDays != null &&
+      (serviceDays.getGraphQLStart() != null || serviceDays.getGraphQLEnd() != null)
+    );
+  }
```

Comment:
This method should be somewhere else, this is not the responsibility of this class and introduce a dependency to the graphQL API classes.

Reply:
Changed in 665ce7ac129049849f5fed1ef05312fe8b874089